File: robocode.api/src/main/java/net/sf/robocode/core/ContainerBase.java
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
package net.sf.robocode.core;


/**
 * @author Pavel Savara (original)
 */
public abstract class ContainerBase {
	public static ContainerBase instance;

	protected abstract <T> T getBaseComponent(java.lang.Class<T> tClass);

	public static <T> T getComponent(java.lang.Class<T> tClass) {
		return instance == null ? null : instance.getBaseComponent(tClass);
	}
}

File: plugins/dotnet/robocode.dotnet.napi/src/net/sf/robocode/core/ContainerBase.cs
using System;

namespace net.sf.robocode.core
{
    public abstract class ContainerBase
    {
        public static ContainerBase instance;

        protected internal abstract object getBaseComponent(Type c);

        private static object getComponentImpl(Type c)
        {
            return ((instance != null) ? instance.getBaseComponent(c) : null);
        }

        public static T getComponent<T>() where T : class
        {
            return (T) getComponentImpl(typeof (T));
        }
    }
}